packages/combo-box/src/vaadin-combo-box-mixin.js

```diff
@@ -782,7 +782,11 @@ export const ComboBoxMixin = (subclass) =>
       // Do not commit value when custom values are disallowed and input value is not a valid option
       // also stop propagation of the event, otherwise the user could submit a form while the input
       // still contains an invalid value
-      if (!this.allowCustomValue && this._inputElementValue !== '' && this._focusedIndex < 0) {
+      const hasInvalidOption =
+        this._focusedIndex < 0 &&
+        this._inputElementValue !== '' &&
+        this._getItemLabel(this.selectedItem) !== this._inputElementValue;
+      if (!this.allowCustomValue && hasInvalidOption) {
```

Comment:
The previous assumption that an invalid/undefined option is detectable by checking for a `_focusedIndex` of `-1` was not correct. Using a data provider, after typing a filter, and then selecting a value, the `filteredItems` are reset, which in turn sets a `_focusedIndex` of `-1`.

The new check additionally tests that the current input value does not match the selected item's label, which would indicate that the user has modified their input, but the `filteredItems` don't contain a matching option.